File: vector.cpp
#include <cmath>

struct Vector
{
    double x, y, z;

    Vector() : x(0.0), y(0.0), z(0.0) {}
    Vector(double x, double y, double z) : x(x), y(y), z(z) {}

    double magnitude() const
    {
        return std::sqrt(x * x + y * y + z * z);
    }

    // Normalizar
    Vector normalize() const
    {
        double m = magnitude();
        if (m > 0.0)
        {
            return Vector(x / m, y / m, z / m);
        }
        return *this;
    }

    // Produto escalar
    double dot(const Vector &other) const
    {
        return x * other.x + y * other.y + z * other.z;
    }

    // Produto vetorial
    Vector cross(const Vector &other) const
    {
        return Vector(y * other.z - z * other.y,
                      z * other.x - x * other.z,
                      x * other.y - y * other.x);
    }

    // Multiplicação por escalar
    Vector scale(double s) const
    {
        return Vector(x * s, y * s, z * s);
    }
};

File: src/vector.py
import math


class Vector:
    def __init__(self, x: float = 0, y: float = 0, z: float = 0):
        self.x: float = x
        self.y: float = y
        self.z: float = z

    def magnitude(self) -> float:
        return math.sqrt(self.x * self.x + self.y * self.y + self.z * self.z)

    def normalize(self) -> 'Vector':
        mag = self.magnitude()
        if mag > 0:
            return Vector(self.x / mag, self.y / mag, self.z / mag)
        return self

    def dot(self, other: 'Vector') -> float:
        return self.x * other.x + self.y * other.y + self.z * other.z

    def cross(self, other: 'Vector') -> 'Vector':
        return Vector(self.y * other.z - self.z * other.y,
                      self.z * other.x - self.x * other.z,
                      self.x * other.y - self.y * other.x)

    def scale(self, scalar: float) -> 'Vector':
        return Vector(self.x * scalar, self.y * scalar, self.z * scalar)

    def subtract(self, other: 'Vector') -> 'Vector':
        return Vector(self.x - other.x, self.y - other.y, self.z - other.z)

    def add(self, other: 'Vector') -> 'Vector':
        return Vector(self.x + other.x, self.y + other.y, self.z + other.z)

    def __str__(self):
        return f"Vector({self.x}, {self.y}, {self.z})"

    def __eq__(self, other):
        if not isinstance(other, Vector):
            return False
        return self.x == other.x and self.y == other.y and self.z == other.z
